### ISSUE TITLE
failed after PEPPER step

### ISSUE THREAD

**Issue by myonaung** [google/deepvariant#497]
Hello,
Thanks very much this amazing software. I have been trying to use Pepper-deepvariant on HPC (under singularity container) with the example dataset you have provided i.e. HG002_guppy_507_2_GRCh38_pass.chr20.30x.bam . However, it failed after PEPPER step "BASIC CANDIDATE FINDER SELECT" and doesn't seem to produce respective vcf as expected. Images generation are completed.

**Comment by MariaNattestad** [comment]
Can you share the error message? Also cc'ing @kishwarshafin who might be better able to help with the PEPPER step.

**Comment by myonaung** [comment]
Thanks for getting back to me.  The error messages are as follow:

Traceback (most recent call last):
  File "/usr/local/bin/pepper_variant", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/pepper_variant/pepper_variant.py", line 70, in main
    call_variant(options)
  File "/usr/local/lib/python3.8/dist-packages/pepper_variant/modules/python/CallVariant.py", line 95, in call_variant
    run_inference(options,
  File "/usr/local/lib/python3.8/dist-packages/pepper_variant/modules/python/RunInference.py", line 141, in run_inference
    distributed_cpu(options,
  File "/usr/local/lib/python3.8/dist-packages/pepper_variant/modules/python/RunInference.py", line 119, in distributed_cpu
    predict_distributed_cpu(options, image_dir, file_chunks, output_dir, callers, threads_per_caller)
  File "/usr/local/lib/python3.8/dist-packages/pepper_variant/modules/python/models/predict_distributed_cpu.py", line 175, in predict_distributed_cpu
    quantize_dynamic(options.model_path + ".onnx", output_filepath + "pepper_model.quantized.onnx", weight_type=QuantType.QUInt8)
  File "/usr/local/lib/python3.8/dist-packages/onnxruntime/quantization/quantize.py", line 279, in quantize_dynamic
    model = load_model(Path(model_input), optimize_model)
  File "/usr/local/lib/python3.8/dist-packages/onnxruntime/quantization/quantize.py", line 48, in load_model
    onnx_model = ONNXModel(optimize_model(Path(model_path)))
  File "/usr/local/lib/python3.8/dist-packages/onnxruntime/quantization/quantize.py", line 40, in optimize_model
    _ = InferenceSession(model_path.as_posix(), sess_option, providers=['CPUExecutionProvider'])
  File "/usr/local/lib/python3.8/dist-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 324, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/usr/local/lib/python3.8/dist-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 358, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
RuntimeError: /onnxruntime_src/onnxruntime/core/platform/posix/env.cc:173 onnxruntime::{anonymous}::PosixThread::PosixThread(const char*, int, unsigned int (*)(int, Eigen::ThreadPoolInterface*), Eigen::ThreadPoolInterface*, const onnxruntime::ThreadOptions&) pthread_setaffinity_np failed, error code: 0 error msg: 


real    6m8.846s
user    5m58.102s
sys     0m8.562s
[12-14-2021 10:20:33] INFO: [3/8] RUNNING THE FOLLOWING COMMAND
-------
mv /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/PEPPER_VARIANT_FULL.vcf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/; 
mv /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/PEPPER_VARIANT_OUTPUT_PEPPER.vcf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/; 
mv /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/PEPPER_VARIANT_OUTPUT_VARIANT_CALLING.vcf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/; 
bgzip /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf; 
tabix -p vcf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz; 
bgzip /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_PEPPER.vcf; 
tabix -p vcf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_PEPPER.vcf.gz; 
bgzip /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_VARIANT_CALLING.vcf; 
tabix -p vcf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_VARIANT_CALLING.vcf.gz; 
rm -rf /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/; 
echo "CONTIGS FOUND IN PEPPER VCF:"; 
zcat /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz | grep -v '#' | cut -f1 | uniq
-------
mv: cannot stat '/data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/PEPPER_VARIANT_FULL.vcf': No such file or directory
mv: cannot stat '/data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/PEPPER_VARIANT_OUTPUT_PEPPER.vcf': No such file or directory
mv: cannot stat '/data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/pepper/PEPPER_VARIANT_OUTPUT_VARIANT_CALLING.vcf': No such file or directory
[bgzip] No such file or directory: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf
tbx_index_build failed: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz
[bgzip] No such file or directory: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_PEPPER.vcf
tbx_index_build failed: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_PEPPER.vcf.gz
[bgzip] No such file or directory: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_VARIANT_CALLING.vcf
tbx_index_build failed: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_OUTPUT_VARIANT_CALLING.vcf.gz
CONTIGS FOUND IN PEPPER VCF:
gzip: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz: No such file or directory
[12-14-2021 10:20:33] INFO: [4/8] RUNNING THE FOLLOWING COMMAND
-------
time margin phase /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//HG002_guppy_507_2_GRCh38_pass.chr20.30x.bam /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//GRCh38_no_alt.chr20.fa /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz /opt/margin_dir/params/phase/allParams.haplotag.ont-r94g507.json -t 1 -V -o /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PHASED.PEPPER_MARGIN 2>&1 | tee /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/logs/2_margin_haplotag.log;
samtools index -@1 /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PHASED.PEPPER_MARGIN.haplotagged.bam
-------
Could not read from vcf file: /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz


real    0m0.069s
user    0m0.001s
sys     0m0.045s
[E::hts_open_format] Failed to open file "/data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PHASED.PEPPER_MARGIN.haplotagged.bam" : No such file or directory
samtools index: failed to open "/data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PHASED.PEPPER_MARGIN.haplotagged.bam": No such file or directory
[12-14-2021 10:20:34] ERROR: None]
[12-14-2021 10:20:34] THE FOLLOWING COMMAND FAILED: time margin phase /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//HG002_guppy_507_2_GRCh38_pass.chr20.30x.bam /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//GRCh38_no_alt.chr20.fa /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PEPPER_VARIANT_FULL.vcf.gz /opt/margin_dir/params/phase/allParams.haplotag.ont-r94g507.json -t 1 -V -o /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/PHASED.PEPPER_MARGIN 2>&1 | tee /data/gpfs/projects/punim1129/probe-cap_seq_data/tools/pepper-deepvariant/mock_data//output/logs/2_margin_haplotag.log;

**Comment by myonaung** [comment]
Thanks for getting back to me. I have attached the screenshot for error messages:
<img width="1368" alt="error_message_deepvariant" src="https://user-images.githubusercontent.com/30123717/145905106-9d351a82-a563-411c-b139-031f0eafa4ad.png">

**Comment by kishwarshafin** [comment]
@myonaung , can you please open this issue in the PEPPER's github repository (https://github.com/kishwarshafin/pepper). I think this issue is with model quantization, your HPC does not support quantized inference.

**Comment by myonaung** [comment]
@kishwarshafin , Thanks. I've opened the issue at Gihtub repository.

**Comment by pgrosu** [comment]
@myonaung I wonder if the error could be something else.  The error seems to indicate that setting the thread to be on a specific CPU is the issue.  Could you try a few things:

1) Can you take this code and put it in a file called `affinity.c` and then compile with `gcc affinity.c -lpthread -o affinity`.   Then run the program `./affinity` and tell us what you see:

```C
#define _GNU_SOURCE
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#define handle_error_en(en, msg) \
       do { errno = en; perror(msg); exit(EXIT_FAILURE); } while (0)

int
main(int argc, char *argv[])
{
   int s;
   cpu_set_t cpuset;
   pthread_t thread;

   thread = pthread_self();

   /* Set affinity mask to include CPUs 0 to 7. */

   CPU_ZERO(&cpuset);
   for (int j = 0; j < 8; j++)
       CPU_SET(j, &cpuset);

   s = pthread_setaffinity_np(thread, sizeof(cpuset), &cpuset);
   if (s != 0)
       handle_error_en(s, "pthread_setaffinity_np");

   /* Check the actual affinity mask assigned to the thread. */

   s = pthread_getaffinity_np(thread, sizeof(cpuset), &cpuset);
   if (s != 0)
       handle_error_en(s, "pthread_getaffinity_np");

   printf("Set returned by pthread_getaffinity_np() contained:\n");
   for (int j = 0; j < CPU_SETSIZE; j++)
       if (CPU_ISSET(j, &cpuset))
           printf("    CPU %d\n", j);

   exit(EXIT_SUCCESS);
}
```

You should see something like this:

```Bash
$ ./affinity
Set returned by pthread_getaffinity_np() contained:
    CPU 0
    CPU 1
    CPU 2
    CPU 3
    CPU 4
    CPU 5
    CPU 6
    CPU 7
$
```

2) Could you tell us what operating system and cpu are your running?  If it is Linux could you the following commands and tell us what you see (these will tell us something about your CPU, Linux version, and total number of threads you can have):

Command 1: `cat /proc/sys/kernel/threads-max`
Command 2: `lscpu`
Command 3: `cat /etc/os-release` 
Command 4: `lsb_release -a`

3) Extra credit:  This will tell us about total number of threads used by the application, and if they might be hitting the maximum allowable thread count. And then just type for the process ID (PID) of the above:

ps -o thcount PID

or 

watch ps -o thcount PID

where PID is the process ID of the process above.  You can get process IDs via the `ps aux` command.

Basically what this will do is tell us if you can actually set affinity on your CPU and the number of cores/threads available.

Thanks,
~p

**Comment by myonaung** [comment]
Hi @pgrosu, thanks very much for the help with troubleshooting. 

1.  affinity.c compilation run into error: 

```
./affinity
pthread_setaffinity_np: Invalid argument.
```

```
2.  cat /proc/sys/kernel/threads-max
2061146
lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                32
On-line CPU(s) list:   0-31
Thread(s) per core:    1
Core(s) per socket:    16
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 63
Model name:            Intel(R) Xeon(R) CPU E5-2698 v3 @ 2.30GHz
Stepping:              2
CPU MHz:               2798.211
CPU max MHz:           3600.0000
CPU min MHz:           1200.0000
BogoMIPS:              4600.13
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              40960K
NUMA node0 CPU(s):     0-15
NUMA node1 CPU(s):     16-31
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm epb invpcid_single intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear spec_ctrl intel_stibp flush_l1d

cat /etc/os-release
NAME="Red Hat Enterprise Linux Server"
VERSION="7.9 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.9"
PRETTY_NAME="Red Hat Enterprise Linux"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.9:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.9
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.9"

lsb_release -a
LSB Version:    :core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
Distributor ID: RedHatEnterpriseServer
Description:    Red Hat Enterprise Linux Server release 7.9 (Maipo)
Release:        7.9
Codename:       Maipo
```

3. I was running it on the interactive mode of HPC with 2 threads.

**Comment by pgrosu** [comment]
@myonaung This tells me you have the hardware to do it, now let's see if the OS has the capability turned on.  Can you please run the following commands:

1) ```cat /proc/filesystems | grep cpu```

You should see something like this:

```Bash
-bash-4.2$ cat /proc/filesystems | grep cpu
nodev   cpuset
-bash-4.2$
```

2) Next let's see if you have these enabled on a process, replacing `USERNAME` with the one for you and the process we use to check for is `bash`:

```ps aux | grep USERNAME | grep bash```

For example, for me it returns this:

```Bash
-bash-4.2$ ps aux | grep pgrosu | grep bash
pgrosu   121490  0.0  0.0 127844  3856 pts/3    Ss   18:25   0:00 -bash
pgrosu   121703  0.0  0.0 112812   976 pts/3    S+   18:40   0:00 grep --color=auto bash
-bash-4.2$
```

Once you have that, then type, replacing `BASH_PID` with the one for you:

```cat /proc/BASH_PID/status | grep allowed```

For me, the BASH_PID is 121490, and when you run the command you should see something like this:

```Bash
-bash-4.2$  cat /proc/121490/status | grep allowed
Cpus_allowed:   ffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff
Cpus_allowed_list:      0-271
Mems_allowed:   00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000003
Mems_allowed_list:      0-1
-bash-4.2$
```

3) Let's see if you have cgroups, by typing this command:

```ls /sys/fs/cgroup/cpuset/```

You should see something like this:

```Bash
-bash-4.2$ ls /sys/fs/cgroup/cpuset/
cgroup.clone_children  cpuset.memory_pressure
cgroup.event_control   cpuset.memory_pressure_enabled
cgroup.procs           cpuset.memory_spread_page
cgroup.sane_behavior   cpuset.memory_spread_slab
cpuset.cpu_exclusive   cpuset.mems
cpuset.cpus            cpuset.sched_load_balance
cpuset.effective_cpus  cpuset.sched_relax_domain_level
cpuset.effective_mems  notify_on_release
cpuset.mem_exclusive   release_agent
cpuset.mem_hardwall    tasks
cpuset.memory_migrate
-bash-4.2$
```

If you do, then try the following commands:

```cat /sys/fs/cgroup/cpuset/cpuset.cpus```
```cat /sys/fs/cgroup/cpuset/cgroup.procs | head -n 7```
```cat /sys/fs/cgroup/cpuset/tasks | head -n 7```

You should see something like this:

```Bash
-bash-4.2$ cat /sys/fs/cgroup/cpuset/cpuset.cpus
0-47
-bash-4.2$  cat /sys/fs/cgroup/cpuset/cgroup.procs | head -n 7
1
2
6
7
8
9
10
-bash-4.2$  cat /sys/fs/cgroup/cpuset/tasks | head -n 7
1
2
6
7
8
9
10
-bash-4.2$
```



Thanks,
~p

**Comment by kishwarshafin** [comment]
@pgrosu ,

Thank you for all the help, however, I found the issue to be in `ONNX` quantization step. I patched and uploaded a new version and provided that to @myonaung. It fixed the issue on their end. As the issue has been resolved, I requested to close this thread. Your query may be unanswered here.
